File: SCOpack/SCOpack/SCOpack/examples/funcLinReg.m
function [out1,out2] = funcLinReg(x,key,T1,T2,A,b)
    % This code provides information for
    %     min   0.5*||Ax-b||^2 
    %     s.t. \|x\|_0<=s
    % where A in R^{m x n} and b in R^{m x 1}    
    switch key
        case 'fg'
            Tx   = find(x~=0);
            Axb  = A(:,Tx)*x(Tx)-b;
            out1 = (Axb'*Axb)/2;      % objective 
            if  nargout == 2 
                out2 = (Axb'*A)';     % gradient 
            end
        case 'h'        
            AT   = A(:,T1); 
            out1 = AT'*AT;            %sub-Hessian formed by rows indexed by T1 and columns indexed by T1   
            if  nargout == 2
                out2 = AT'*A(:,T2);   %sub-Hessian formed by rows indexed by T1 and columns indexed by T2
            end       
    end
end

File: SCOpack/SCOpack/SCOpack/examples/funcLinReg.py
import numpy as np
from typing import Tuple, Optional


def funcLinReg(x: np.ndarray, key: str, T1: Optional[np.ndarray], T2: Optional[np.ndarray], A: np.ndarray, b: np.ndarray) -> Tuple[np.ndarray, Optional[np.ndarray]]:
    """
    Python port of funcLinReg.m
      Problem: min 0.5*||A x - b||^2 subject to ||x||_0 <= s
    Returns (out1,) for key=='fg' or 'h'; and (out1, out2) if gradient/sub-Hessian also requested.
    """
    x = np.asarray(x).reshape(-1)
    if key == 'fg':
        Tx = np.flatnonzero(x)
        Axb = (A[:, Tx] @ x[Tx]) - b
        out1 = 0.5 * float(Axb.T @ Axb)
        # Gradient if needed
        grad = A.T @ Axb
        return out1, grad
    elif key == 'h':
        T1 = np.asarray(T1, dtype=int)
        AT = A[:, T1]
        H11 = AT.T @ AT
        if T2 is None:
            return H11
        T2 = np.asarray(T2, dtype=int)
        H12 = AT.T @ A[:, T2]
        return H11, H12
    else:
        raise ValueError("key must be 'fg' or 'h'")
